src/lib/token_tree.js

```diff
@@ -0,0 +1,92 @@
+export default class TokenTree {
+  constructor() {
+    this.rootNode = { root: true, children: [] };
+    this.stack = [ this.rootNode ];
+  }
+
+  get top() {
+    return this.stack[this.stack.length - 1];
+  }
+
+  add(node) {
+    this.top.children.push(node);
+  }
+
+  addKeyword(text, kind) {
+    if (text === "") { return; }
+
+    this.top.children.push({
+      kind,
+      children: [ text ]
+    });
+  }
+
+  addText(text) {
+    if (text === "") { return; }
+
+    this.add(text);
+  }
+
+  addSublanguage({rootNode}, name) {
+    let node = rootNode;
+    delete node.root; // no longer a root node
+    node.kind = name;
+    node.sublanguage = true;
+    this.add(node);
+  }
+
+  openNode(kind) {
+    let node = { kind, children: [] };
+    this.add(node);
+    this.stack.push(node);
+  }
+
+  closeNode() {
+    if (this.stack.length > 1)
+      this.stack.pop();
+  }
+
+  closeAllNodes() {
+    while (this.stack.length > 1)
+      this.closeNode();
+  }
```

Comment:
Just a matter of style, but for me `while (this.closeNode());` would look better (to not repeat the same check).

Reply:
This requires closeNode now return a value, which really it's now only doing to help out closeAllNodes, which has a certain smell for me, but I don't care that much, so I'll change.